File: syn/design/syn/AdapterLib.cs
namespace syn
{
    using System.Collections.Generic;

    public static class AdapterLib
    {
        // typedef List<AdapterInfo> (WINAPI* ALGAI)();
        public delegate List<AdapterInfo> ALGAI();

        private static List<IAdapter> adapters;

        public static List<AdapterInfo> GetAdapterInfos()
        {
            var adapterInfos = new List<AdapterInfo>();
            foreach (var adapter in adapters)
            {
                adapterInfos.Add(adapter.GetInfo());
            }
            return adapterInfos;
        }

        public static int DllMain()
        {
            // prepare adapters list
            adapters.Add(new SynthAdapter());
            return 0;
        }
    }
}

File: syn/test/adapterlib/adapterlib.cpp
#include "adapterlib.h"
#include "base/memory.h"
#include "collection/array.h"
#include "ssn1k/synthadapter.h"

NS_FW_BASE_USE
NS_PLAYER_USE
NS_SSN1K_USE

//void addAdapter(IAdapter* adapter) {
//	AdapterLib::addAdapter(adapter);
//}

IAdapter* getAdapter(int uid) {
	return AdapterLib::getAdapter(uid);
}

Map& getAdapters() {
	return AdapterLib::getAdapters();
}

BOOL WINAPI DllMain(
	_In_ HINSTANCE hinstDLL,
	_In_ DWORD     fdwReason,
	_In_ LPVOID    lpvReserved
) {
	switch (fdwReason)
	{
		case DLL_PROCESS_ATTACH:
			// Initialize once for each new process.
			// Return FALSE to fail DLL load.
			// prepare adapters list
			AdapterLib::addAdapter(NEW_(SynthAdapter));
			break;

		case DLL_THREAD_ATTACH:
			// Do thread-specific initialization.
			break;

		case DLL_THREAD_DETACH:
			// Do thread-specific cleanup.
			break;

		case DLL_PROCESS_DETACH:
			// Perform any necessary cleanup.
			break;
	}
	return TRUE;
}
